ramips:  add uboot-envtools to mt7621 DEFAULT_PACKAGES

jonathan-sturges opened the pull request:
ramips: fix installation issues for Amped Wireless ALLY
    
add uboot-envtools to mt7621 DEFAULT_PACKAGES to support
two-stage initial installation, but remove it via DEVICE_PACKAGES
for devices not needing it.

Signed-off-by: Jonathan Sturges <jsturges@redhat.com>

mkresin commented on target/linux/ramips/mt7621/base-files/lib/upgrade/platform.sh:
```diff
@@ -29,8 +29,10 @@ platform_do_upgrade() {
 	ampedwireless,ally-r1900k)
-		if [ "$(fw_printenv --lock / -n bootImage 2>/dev/null)" != "0" ]; then
-			fw_setenv --lock / bootImage 0 || exit 1
+		if [ "$(grep -c bootImage=0 /dev/mtd1)" != "1" ]; then
```
Please add a note to the commit message why you want to get rid of the uboot-envtools dependency.

Beside if that, it doesn't seem to me a good idea to do so. I mean, the uboot-envtools know about the inner structure of the u-boot env. With the change you are switching to hackish way of changing parameters stored in the uboot-env instead.

Furthermore, I'm in doubt that it works at all. To my knowledge, the u-boot env is protected by a checksum. After changing the bootImage parameter and without updating the checksum, u-boot should detect the env as corrupt and should use its internal env instead. U-boot should print a big warning about the corrupted env during start.

jonathan-sturges replied:
I would love to continue requiring uboot-envtools, but as of 21.02.x, it is not included in the release initramfs images despite being listed in DEVICE_PACKAGES.  The ALLY devices require a 2-stage initial installation (initramfs first, then sysupgrade) and require a way to check and/or set the boot partition for sysupgrade.  fw_printenv/fw_setenv would be great but the dd/sed/mtd workaround works fine in my testing.

jonathan-sturges replied:
You were correct, it detects the env as corrupted.  However it still works, because it reverts to the default environment, which uses the first boot image anyway.  So, just lucky there.
How do you recommend proceeding?  Is there a way I can mandate that uboot-envtools be embedded in these devices' initramfs?  Thanks.

hauke replied:
For initramfs builds the same rootfs will be used for all devices. You can add `uboot-envtools` to the `DEFAULT_PACKAGES` in target/linux/ramips/mt7621/target.mk and remove it from all `DEVICE_PACKAGES` for all mt7621 devices which do not use it with `-DEVICE_PACKAGES` Then it should be integrated in the default initramfs images.

jonathan-sturges replied:
Thanks for the idea, I'll test this out.  It sounds fine to me as long as it doesn't end up increasing the initramfs image size too much for other mt7621 devices.

jonathan-sturges replied:
This works in my testing, thanks.

DragonBluep replied:
don't know what's going on with 21.02, but at least on the master branch uboot-envtools can be integrated into the initramfs. I just used it a few months ago. see 13308161788c98ae6cd48c22b13339fdb8c77130

DragonBluep replied:
Strange, comparing the official initramfs and sysupgrade binary sizes, it seems that even the WiFi driver is not included.😳

hauke commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -128,14 +128,15 @@ TARGET_DEVICES += alfa-network_quad-e4g
   PAGESIZE := 2048
   UBINIZE_OPTS := -E 5
-  KERNEL_INITRAMFS := $(KERNEL_DTB) | uImage lzma -n 'flashable-initramfs' |\
+  KERNEL_INITRAMFS := $(KERNEL_DTB) | loader-kernel | lzma | uImage lzma |\
```
Please add the `uimage-lzma-loader` changes into a separate commit.

jonathan-sturges replied:
I was thinking about that too; done.

jonathan-sturges committed:
ramips: fix installation issues for Amped Wireless ALLY

add uboot-envtools to mt7621 DEFAULT_PACKAGES to support
two-stage initial installation, but remove it via DEVICE_PACKAGES
for devices not needing it.

Signed-off-by: Jonathan Sturges <jsturges@redhat.com>

Ansuel commented:
I'm not following this... Why? what is the issue? Also do we really need all this delta... is it worth it?

jonathan-sturges commented:
I agree the delta is a bit ugly but I feel it's with good reason:  many mt7621 devices need uboot-envtools in the initramfs in order to do a two-step OpenWRT installation.  For example the Amped Wireless devices have multiple flash regions and the initramfs needs to have the tools to check & set the proper boot region in platorm.sh.

Ansuel commented:
yes then just lets add uboot-envtools for every package... the image increase can't be that big... also these device are not space-constraints right?

jonathan-sturges commented:
The issue (as I understand it) is DEVICE_PACKAGES are not included in initramfs images, just the DEFAULT_PACKAGES.  Open to other solutions though.

Ansuel commented:
@jonathan-sturges i would just put uboot-envtool in DEFAULT_PACKAGES and on another commit drop the uboot-envtool in every device adding it in DEVICE-PACKAGES

Simple change and the increase it's really 10kb

Spudz76 commented:
My linksys ea7300-v2 already builds and includes it and so I'm confused.  Also a dual-bank device that needs to set vars for active boot partition.
```
CONFIG_DEFAULT_uboot-envtools=y
CONFIG_PACKAGE_uboot-envtools=y
```

jonathan-sturges commented:
@Ansuel OK will do.

Djfe commented:
A similar patch was merged on March 1st.
If there are still changes that you require, then please rebase. Else you might want to close this PR :)
https://github.com/openwrt/openwrt/commit/79bd0172ca5671f7cdf382dc80ec1630da057947
